scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala

```diff
@@ -0,0 +1,162 @@
+    var res: List[ScParameter] = Nil
+    anonClass.accept(new ScalaRecursiveElementVisitor() {
+      override def visitReferenceExpression(expression: ScReferenceExpression) = {
+        val refElement = expression.resolve
+        refElement match {
+          case p: ScParameter =>
+            val containingClass = PsiTreeUtil.getParentOfType(p, classOf[ScClass])
+            val containingObject = PsiTreeUtil.getParentOfType(p, classOf[ScObject])
+            val isTopLevelScala3Definition = containingObject == null && containingClass == null
+            if (PsiTreeUtil.isAncestor(containingClass, anonClass, true) || PsiTreeUtil.isAncestor(containingObject, anonClass, true) || isTopLevelScala3Definition) {
+              res = p :: res
+            }
+          case _ =>
+        }
+        super.visitReferenceExpression(expression)
+      }
+    })
+    res.reverse.distinct.map(parameter =>
+      new ScalaVariableData(parameter, true, parameter.`type`().getOrNothing)
+    ).toArray
+
+  }
+
+  private case class DialogResult(className: String, variables: Array[VariableData])
+
+  private def showRefactoringDialog(project: Project, anonClass: ScExtendsBlock, usedVariables: Array[VariableData], target: Either[ScFile, ScTemplateDefinition]) = {
+    val dialog = new ScalaAnonymousToInnerDialog(project, anonClass, usedVariables, target)
+    if (dialog.showAndGet()) {
+      Some(DialogResult(dialog.getClassName, dialog.getVariables))
+    } else
+      None
+  }
+
+  private def createClass(name: String, anonClass: ScExtendsBlock, variables: Array[VariableData], project: Project): ScClass = {
+    implicit val projectContext: ProjectContext = new ProjectContext(project)
+
+    val parametersText = variables.map(v => s"${v.name}: ${v.`type`.getPresentableText}").mkString(", ")
+    val parameterOldNameToNewName = variables.map(v => v.variable.getName -> v.name).toMap
+
+    val text = s"class $name($parametersText) extends ${anonClass.getText}"
+
+    val newClass = createElementFromText[ScClass](text, anonClass)
+
+    newClass.accept(new ScalaRecursiveElementVisitor() {
+      override def visitReferenceExpression(expression: ScReferenceExpression) = {
+        val refElement = expression.resolve
+        refElement match {
+          case p: ScParameter =>
+            expression.handleElementRename(parameterOldNameToNewName(p.getName))
+          case _ =>
+        }
+        super.visitReferenceExpression(expression)
+      }
+    })
+
+    newClass
+  }
+
+  @tailrec
+  private def findAnonymousClass(element: PsiElement): Option[ScExtendsBlock] =
+    Option(element) match {
+      case Some(n: ScNewTemplateDefinition) =>
+        n.findFirstChildByTypeScala[ScExtendsBlock](ScalaElementType.EXTENDS_BLOCK)
+      case Some(_) => findAnonymousClass(element.getParent)
+      case None => None
+    }
+
+  @tailrec
+  private def findTargetContainer(elem: PsiElement): Either[ScFile, ScTemplateDefinition] = {
```

Comment:
I'll add `ScTrait`. but `ScEnum` is redundant because it extends `ScClass` anyway

Reply:
Could you please at least add a small meaningful test for ScEnum case (and ScTrait)
The fact that `ScEnum` extends `ScClass` is an implementation detail.
We had a lot of refactorings of `ScEnum` in 2023.2/3 and in future there is still a chance the hierarchy will be remade.

Reply:
I have added a test for enum but there's some issue in the testkit where it creates the virtual file for the test from the string it doesn't build the AST correctly and doesn't contain the `ScEnum` in it. manually running the function when intelliJ is running works correctly.

Since the test doesn't create the ScEnum node correctly, the test fails because the created class is generated on the top level instead of inside the enum. please try to run `ScalaAnonymousToInnerTest.scala#testInsideTrait`.
The rest of the tests pass correctly.

Reply:
Consider overriding `supportedIn` method in the test class.
Example:
```scala
override protected def supportedIn(version: ScalaVersion): Boolean = version.isScala3
```